File: cpp/lib/1008.ConstructBinarySearchTreeFromPreorderTraversal.cpp
// 1008. Construct Binary Search Tree from Preorder Traversal

#include "1008.ConstructBinarySearchTreeFromPreorderTraversal.h"


TreeNode* LeetCode1008ConstructBinarySearchTreeFromPreorderTraversal::GenBinarySearchTree_(vector<int>& preorder, int leftIndex, int rightIndex) {
	if (leftIndex > rightIndex) return NULL;

	auto val = preorder[leftIndex];
	auto node = new TreeNode(val);
	if (leftIndex == rightIndex) return node;

	auto seperator = rightIndex+1;
	for (int i = leftIndex+1; i <= rightIndex; ++i) {
		if (preorder[i] > val) {
			seperator = i;
			break;
		}
	}
	node->left = GenBinarySearchTree_(preorder, leftIndex+1, seperator-1);
	node->right = GenBinarySearchTree_(preorder, seperator, rightIndex);
	return node;
}

TreeNode* LeetCode1008ConstructBinarySearchTreeFromPreorderTraversal::bstFromPreorder(vector<int>& preorder) {
	if (preorder.size() == 0) return NULL;
	return GenBinarySearchTree_(preorder, 0, preorder.size() - 1);
}

File: csharp/src/1008_ConstructBinarySearchTreeFromPreorderTraversal.cs
/*
LeetCode: 1008. Construct Binary Search Tree from Preorder Traversal
*/
using System.Linq;

namespace LeetCode.Problem_1008
{    
	public class ConstructBinarySearchTreeFromPreorderTraversal {
		public class TreeNode {
			public int val;
			public TreeNode left;
			public TreeNode right;
			public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
				this.val = val;
				this.left = left;
				this.right = right;
			}
		}

		public TreeNode BstFromPreorder(int[] preorder) {
			return _BuildTree(preorder, 0, preorder.Length-1);
		}

		private TreeNode _BuildTree(int[] preorder, int left, int right)
		{
			if (left > right) return null;

			var val = preorder[left];
			var node = new TreeNode(val);

			var index = left;
			for (; index <= right; ++index)
				if (preorder[index] > val) break;
			
			node.left = _BuildTree(preorder, left+1, index-1);
			node.right = _BuildTree(preorder, index, right);
			return node;
		}
	}
}
